api/src/services/opportunities_v1/opportunity_version.py

```diff
@@ -11,23 +14,34 @@
 
 def save_opportunity_version(db_session: db.Session, opportunity: Opportunity) -> None:
     """
-    Saves a new version of an Opportunity record in the OpportunityVersion table.
+    Saves a new version of an Opportunity record in the OpportunityVersion table if there are changes.
 
-    This function extracts the opportunity data from the provided Opportunity model instance,
-    creates a new OpportunityVersion record with the data, and saves it in the database.
+    This function first fetches the most recent version of the Opportunity record stored in the OpportunityVersion table.
+    It compares the existing data with the current Opportunity record, and if there are any differences, it creates
+    a new OpportunityVersion record and saves it in the database in a JSON format.
 
     :param  db_session: The active SQLAlchemy session used to interact with the database.
     :param opportunity: An instance of the Opportunity model containing the data to be saved.
     :return: This function does not return a value. It saves a new version of the opportunity in the database.
     """
-    #  Extracts the opportunity data as JSON object
-
-    schema_data = SCHEMA.dump(opportunity)
-
-    # Add new OpportunityVersion instance to the database session
-    opportunity_version = OpportunityVersion(
-        opportunity_id=opportunity.opportunity_id,
-        opportunity_data=schema_data,
-    )
 
-    db_session.add(opportunity_version)
+    # Fetch latest opportunity version stored
+    latest_opp_version = db_session.execute(
+        select(OpportunityVersion)
+        .where(OpportunityVersion.opportunity_id == opportunity.opportunity_id)
+        .order_by(OpportunityVersion.created_at.desc())
+    ).scalar_one_or_none()
```

Comment:
One additional recommendation:
```suggestion
    latest_opp_version = db_session.execute(
        select(OpportunityVersion)
        .where(OpportunityVersion.opportunity_id == opportunity.opportunity_id)
        .order_by(OpportunityVersion.created_at.desc())
        .options(selectinload("*"))
    ).scalar_one_or_none()
```

Won't have any obvious affect on how the code works, but makes it so SQLAlchemy will select all parts of an opportunity in one set of queries rather than lazy-loading every relationship which each individually fire a select query off to the database. When we have a job that'll iterate over tens of thousands of records, it'll make it much faster (I tested adding it with set-current-opportunities and I want to say it was 70x faster?)

Reply:
Yes indeed! Updated , thanks